File: app/src/main/java/wangdaye/com/geometricweather/common/ui/adapters/AnimationAdapterWrapper.java
package wangdaye.com.geometricweather.common.ui.adapters;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.HashMap;
import java.util.Map;

public abstract class AnimationAdapterWrapper<A extends RecyclerView.Adapter<VH>, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> {

    private final A mInner;
    private final Map<Integer, Animator> mAnimatorSet;

    private int mLastPosition;
    private boolean mFirstOnly;

    public AnimationAdapterWrapper(A adapter) {
        this(adapter, true);
    }

    public AnimationAdapterWrapper(A adapter, boolean firstOnly) {
        super.setHasStableIds(adapter.hasStableIds());

        mInner = adapter;
        mAnimatorSet = new HashMap<>();

        mLastPosition = -1;
        mFirstOnly = firstOnly;
    }

    @Nullable
    protected abstract Animator getAnimator(View view, int pendingCount);

    protected abstract void setInitState(View view);

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return mInner.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void registerAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        super.registerAdapterDataObserver(observer);
        mInner.registerAdapterDataObserver(observer);
    }

    @Override
    public void unregisterAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        super.unregisterAdapterDataObserver(observer);
        mInner.unregisterAdapterDataObserver(observer);
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        mInner.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        mInner.onDetachedFromRecyclerView(recyclerView);
    }

    @Override
    public void onViewAttachedToWindow(@NonNull VH holder) {
        super.onViewAttachedToWindow(holder);
        mInner.onViewAttachedToWindow(holder);
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull VH holder) {
        super.onViewDetachedFromWindow(holder);
        mInner.onViewDetachedFromWindow(holder);
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        mInner.onBindViewHolder(holder, position);

        if (!mFirstOnly || position > mLastPosition) {
            clear(holder.itemView, position);

            Animator a = getAnimator(holder.itemView, mAnimatorSet.size());
            if (a != null) {
                setInitState(holder.itemView);

                a.addListener(new AnimatorListenerAdapter() {

                    @Override
                    public void onAnimationEnd(Animator animation) {
                        mAnimatorSet.remove(position);
                    }
                });
                a.start();

                mAnimatorSet.put(position, a);
                mLastPosition = position;
                return;
            }
        }
        clear(holder.itemView, position);
    }

    private void clear(View view, int position) {
        Animator a = mAnimatorSet.get(position);
        if (a != null) {
            a.cancel();
            mAnimatorSet.remove(position);
        }

        view.setAlpha(1f);

        view.setRotation(0f);
        view.setRotationX(0f);
        view.setRotationY(0f);

        view.setScaleX(1f);
        view.setScaleY(1f);
        view.setTranslationX(0f);
        view.setTranslationY(0f);
        view.setTranslationZ(0f);
    }

    @Override
    public void onViewRecycled(@NonNull VH holder) {
        super.onViewRecycled(holder);
        mInner.onViewRecycled(holder);
        clear(holder.itemView, holder.getBindingAdapterPosition());
    }

    @Override
    public int getItemCount() {
        return mInner.getItemCount();
    }

    @Override
    public int getItemViewType(int position) {
        return mInner.getItemViewType(position);
    }

    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(hasStableIds);
        mInner.setHasStableIds(hasStableIds);
    }

    @Override
    public long getItemId(int position) {
        return mInner.getItemId(position);
    }

    public A getWrappedAdapter() {
        return mInner;
    }

    public void setLastPosition(int lastPosition) {
        mLastPosition = lastPosition;
    }

    public void setFirstOnly(boolean firstOnly) {
        mFirstOnly = firstOnly;
    }
}

File: app/src/main/java/org/breezyweather/common/ui/adapters/AnimationAdapterWrapper.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.ui.adapters

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver

abstract class AnimationAdapterWrapper<A : RecyclerView.Adapter<VH>, VH : RecyclerView.ViewHolder>
@JvmOverloads constructor(
    adapter: A,
    firstOnly: Boolean = true,
) : RecyclerView.Adapter<VH>() {
    val wrappedAdapter: A
    private val mAnimatorSet: MutableMap<Int, Animator>
    private var mLastPosition: Int
    private var mFirstOnly: Boolean

    init {
        super.setHasStableIds(adapter.hasStableIds())
        wrappedAdapter = adapter
        mAnimatorSet = HashMap()
        mLastPosition = -1
        mFirstOnly = firstOnly
    }

    protected abstract fun getAnimator(view: View, pendingCount: Int): Animator?
    protected abstract fun setInitState(view: View)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        return wrappedAdapter.onCreateViewHolder(parent, viewType)
    }

    override fun registerAdapterDataObserver(observer: AdapterDataObserver) {
        super.registerAdapterDataObserver(observer)
        wrappedAdapter.registerAdapterDataObserver(observer)
    }

    override fun unregisterAdapterDataObserver(observer: AdapterDataObserver) {
        super.unregisterAdapterDataObserver(observer)
        wrappedAdapter.unregisterAdapterDataObserver(observer)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        wrappedAdapter.onAttachedToRecyclerView(recyclerView)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        wrappedAdapter.onDetachedFromRecyclerView(recyclerView)
    }

    override fun onViewAttachedToWindow(holder: VH) {
        super.onViewAttachedToWindow(holder)
        wrappedAdapter.onViewAttachedToWindow(holder)
    }

    override fun onViewDetachedFromWindow(holder: VH) {
        super.onViewDetachedFromWindow(holder)
        wrappedAdapter.onViewDetachedFromWindow(holder)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        wrappedAdapter.onBindViewHolder(holder, position)
        if (!mFirstOnly || position > mLastPosition) {
            clear(holder.itemView, position)
            val a = getAnimator(holder.itemView, mAnimatorSet.size)
            if (a != null) {
                setInitState(holder.itemView)
                a.addListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        mAnimatorSet.remove(position)
                    }
                })
                a.start()
                mAnimatorSet[position] = a
                mLastPosition = position
                return
            }
        }
        clear(holder.itemView, position)
    }

    private fun clear(view: View, position: Int) {
        val a = mAnimatorSet[position]
        if (a != null) {
            a.cancel()
            mAnimatorSet.remove(position)
        }
        view.apply {
            alpha = 1f
            rotation = 0f
            rotationX = 0f
            rotationY = 0f
            scaleX = 1f
            scaleY = 1f
            translationX = 0f
            translationY = 0f
            translationZ = 0f
        }
    }

    override fun onViewRecycled(holder: VH) {
        super.onViewRecycled(holder)
        wrappedAdapter.onViewRecycled(holder)
        clear(holder.itemView, holder.bindingAdapterPosition)
    }

    override fun getItemCount(): Int {
        return wrappedAdapter.itemCount
    }

    override fun getItemViewType(position: Int): Int {
        return wrappedAdapter.getItemViewType(position)
    }

    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(hasStableIds)
        wrappedAdapter.setHasStableIds(hasStableIds)
    }

    override fun getItemId(position: Int): Long {
        return wrappedAdapter.getItemId(position)
    }

    fun setLastPosition(lastPosition: Int) {
        mLastPosition = lastPosition
    }

    fun setFirstOnly(firstOnly: Boolean) {
        mFirstOnly = firstOnly
    }
}
